stalwart-mail: Fix spam-filter missing from `/etc`

norpol opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

* [x] add norpol to nixpkgs maintainers
* [x] chore work with nixpkgs style preferences in other sub packages
* [x] The current nix unstable package doesn't contain the `spamfilter.toml` anylonger. This pull requests packages the spam-filter as an additional package.


```
$ find /nix/store/jwznrq528mx0z844ka6j4rfqk5bhj463-stalwart-mail-0.12.4/
/nix/store/jwznrq528mx0z844ka6j4rfqk5bhj463-stalwart-mail-0.12.4/
/nix/store/jwznrq528mx0z844ka6j4rfqk5bhj463-stalwart-mail-0.12.4/bin
/nix/store/jwznrq528mx0z844ka6j4rfqk5bhj463-stalwart-mail-0.12.4/bin/stalwart-cli
/nix/store/jwznrq528mx0z844ka6j4rfqk5bhj463-stalwart-mail-0.12.4/bin/stalwart
/nix/store/jwznrq528mx0z844ka6j4rfqk5bhj463-stalwart-mail-0.12.4/lib
/nix/store/jwznrq528mx0z844ka6j4rfqk5bhj463-stalwart-mail-0.12.4/lib/systemd
/nix/store/jwznrq528mx0z844ka6j4rfqk5bhj463-stalwart-mail-0.12.4/lib/systemd/system
/nix/store/jwznrq528mx0z844ka6j4rfqk5bhj463-stalwart-mail-0.12.4/lib/systemd/system/stalwart-mail.service
/nix/store/jwznrq528mx0z844ka6j4rfqk5bhj463-stalwart-mail-0.12.4/etc
/nix/store/jwznrq528mx0z844ka6j4rfqk5bhj463-stalwart-mail-0.12.4/etc/stalwart
```

Note: Stalwart will still download additional data through the `spam-filter.toml`. I don't think this has changed across the releases. I don't see that the ASN IPs, geolite, domains_mx, free_email_providers, ... are feasible to package into nixpkgs. So I'm not even sure whether it is advisable to not pull the spam-filters automatically from GitHub as well (so staying with the upstream config).

Also I'm unsure whether the auto-update flag has only an impact on the filter rules or also the http URLs downloaded through the filters.

```
$ grep 'https://'  /nix/store/ij1msx3awgpajyk4ci3zws3pyyixawkf-spam-filter-2.0.3/spam-filter.toml
asn = [ "https://cdn.jsdelivr.net/npm/@ip-location-db/asn/asn-ipv4.csv",
        "https://cdn.jsdelivr.net/npm/@ip-location-db/asn/asn-ipv6.csv" ]
geo = [ "https://cdn.jsdelivr.net/npm/@ip-location-db/geolite2-geo-whois-asn-country/geolite2-geo-whois-asn-country-ipv4.csv",
        "https://cdn.jsdelivr.net/npm/@ip-location-db/geolite2-geo-whois-asn-country/geolite2-geo-whois-asn-country-ipv4.csv" ]
url = "https://openphish.com/feed.txt"
url = "https://disposable.github.io/disposable-email-domains/domains_mx.txt"
url = "https://gist.githubusercontent.com/okutbay/5b4974b70673dfdcc21c517632c1f984/raw/993a35930a8d24a1faab1b988d19d38d92afbba4/free_email_provider_domains.txt"
```

built generates the same result
```
sha256sum /nix/store/ij1msx3awgpajyk4ci3zws3pyyixawkf-spam-filter-2.0.3/spam-filter.toml <(curl -qs -L -o - https://github.com/stalwartlabs/spam-filter/releases/download/v2.0.3/spam-filter.toml)
1b614bc004284116f6c98b5dfb2ce663872b5f60a0a8782c06ded9dcf3da2cd2  /nix/store/ij1msx3awgpajyk4ci3zws3pyyixawkf-spam-filter-2.0.3/spam-filter.toml
1b614bc004284116f6c98b5dfb2ce663872b5f60a0a8782c06ded9dcf3da2cd2  /proc/self/fd/11
```


* 

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

norpol committed:
stalwart-mail: Fix spam-filter missing from `/etc`

norpol commented:
FYI: @diogotcorreia I've seen you've adjusted the spam-filter topic in #412054. I can't pin-point the exact commit, but stalwart isn't shipping the rules through the Rust build any longer.

norpol commented on nixos/modules/services/mail/stalwart-mail.nix:
```diff
@@ -107,7 +107,8 @@ in
       ];
-      spam-filter.resource = lib.mkDefault "file://${cfg.package}/etc/stalwart/spamfilter.toml";
+      spam-filter.resource = lib.mkDefault "file://${cfg.package.spam-filter}/spam-filter.toml";
+      spam-filter.auto-update = lib.mkDefault false;
```
not sure whether this is advisable, I don't know how stalwart is downloading the `https://` URLs inside the file itself.

norpol replied:
https://github.com/stalwartlabs/stalwart/discussions/1770

I've asked a question in the discussion channel.

norpol commented on pkgs/by-name/st/stalwart-mail/spam-filter.nix:
```diff
@@ -0,0 +1,40 @@
+    description = "Spam Filter Rules";
+    homepage = "https://github.com/stalwartlabs/spam-filter";
+    changelog = "https://github.com/stalwartlabs/spam-filter/blob/${src.tag}/CHANGELOG.md";
+    license = lib.licenses.mit;
```
On GitHub they have both an Apache as an MIT License. I'm not sure how you would pass that through the meta tag.

provokateurin replied:
You can use a list and put both licenses in there.

norpol replied:
Thank you :)

diogotcorreia commented:
@norpol Seems to have been changed in this commit upstream: https://github.com/stalwartlabs/stalwart/commit/38fa0291e2f3a47699a77388b5aa8e3eca4f660b

diogotcorreia commented on pkgs/by-name/st/stalwart-mail/spam-filter.nix:
```diff
@@ -0,0 +1,43 @@
+  src = fetchFromGitHub {
+    owner = "stalwartlabs";
+    repo = "spam-filter";
+    tag = "v${version}";
```
```suggestion
    tag = "v${finalAttrs.version}";
```

diogotcorreia commented on pkgs/by-name/st/stalwart-mail/spam-filter.nix:
```diff
@@ -0,0 +1,43 @@
+  nix-update-script,
+}:
+
+stdenv.mkDerivation rec {
```
```suggestion
stdenv.mkDerivation (finalAttrs: {
```

Use the `finalAttrs` pattern instead (see https://github.com/NixOS/nixpkgs/pull/119942)

norpol replied:
TIL, should I do those changes for all other stalwart-mail packages too then? I was just following the webadmin setup.

diogotcorreia commented on pkgs/by-name/st/stalwart-mail/spam-filter.nix:
```diff
@@ -0,0 +1,43 @@
+  meta = {
+    description = "Spam Filter Rules";
+    homepage = "https://github.com/stalwartlabs/spam-filter";
+    changelog = "https://github.com/stalwartlabs/spam-filter/blob/${src.tag}/CHANGELOG.md";
```
```suggestion
    changelog = "https://github.com/stalwartlabs/spam-filter/blob/${finalAttrs.src.tag}/CHANGELOG.md";
```

norpol commented on pkgs/by-name/st/stalwart-mail/spam-filter.nix:
```diff
@@ -0,0 +1,43 @@
+  };
+
+  meta = {
+    description = "Spam Filter Rules";
```
Thanks. I'll incorporate your suggestions.

Concerning naming:
https://github.com/stalwartlabs/stalwart/commit/9cf2863a8e39bf6f386f17597849d8e842a7d137
> "Stalwart Mail Server is now just Stalwart"

I would guess they're possibly going to update the spam-filters README.

norpol commented on pkgs/by-name/st/stalwart-mail/spam-filter.nix:
```diff
@@ -0,0 +1,43 @@
+      mit
+      asl20
+    ];
+    maintainers = with lib.maintainers; [ norpol ];
```
Commited this + adjusted the webmail sub-package to do the same.

norpol committed:
stalwart-mail: webmail, ensure overridable recursive attributes

Follows https://github.com/NixOS/nixpkgs/pull/119942

norpol committed:
maintainers: add norpol

diogotcorreia commented:
Any reason for including the update to 0.12.5 here instead of on a separate PR?
As for the enterprise feature problem, we should definitely not enable the feature by default, and should instead apply a patch with this commit: https://github.com/stalwartlabs/stalwart/commit/861471fde4285b35c6ccc4720730e630ff84563c

norpol commented:
@diogotcorreia Just wanted to give the build a try, since there is plenty of feedback on the current PR I've switched it back to a draft since I'd expect more rounds. From the GitHub PRs in general it's not uncommon that one PR is addressing multiple things - although that makes back-porting sometimes harder. Just added the `.patch` with the upstream fix to see how it goes.

diogotcorreia commented on pkgs/by-name/st/stalwart-mail/spam-filter.nix:
```diff
@@ -0,0 +1,43 @@
+    hash = "sha256-NhD/qUiGhgESwR2IOzAHfDATRlgWMcCktlktvVfDONk=";
+  };
+
+  # TODO Potentially retrieve / pre-fetch / store `https://` resources inside the file
```
@norpol might make sense to remove the comment if we agree we shouldn't interfere with the https:// resources inside the file; your call

norpol replied:
Sorry, overlooked this one - I've removed the comment now. I'm not sure if there is a big benefit of not automatically sourcing the spam-filter through https then as well, kinda packaging half of it through nix feels imperfect - but for now this at least restores the previous intent of the other nix package maintainers.

diogotcorreia commented on pkgs/by-name/st/stalwart-mail/webadmin.nix:
```diff
@@ -14,65 +15,67 @@
-  pname = "webadmin";
-  version = "0.1.27";
+rustPlatform.buildRustPackage (
+  finalAttrs: with finalAttrs; {
```
nit; I'd avoid the use of `with finalAttrs`, I think it makes it less clear what attributes are in scope and where they come from. Just use `finalAttrs.<attr>` instead, there aren't that many uses I believe (and also avoids the massive whitespace change I think)

diogotcorreia commented on pkgs/by-name/st/stalwart-mail/package.nix:
```diff
@@ -20,17 +21,26 @@
+  cargoHash = "sha256-62aPJdqiw7+9ilmULCPuGtfzembxKYwuZyGrgQX3xCE=";
+
+  # Add workaround for building OSS variant for 0.12.5
+  # https://github.com/stalwartlabs/stalwart/commit/861471fde4285b35c6ccc4720730e630ff84563c.patch
```
Change this comment to link to the relevant issue instead (the patch is already visible below, so no need)

https://github.com/stalwartlabs/stalwart/issues/1724

diogotcorreia reviewed (commented):
A few more nitpicks:

- add "Closes <pr>" for both the existing PRs that update stalwart (make sure to add "Closes" before both, otherwise they won't be closed)
- you probably should squash some of the commits (e.g., the ones fixing the spam-filter package should just be merged into the first some)
- the commit adding yourself to `maintainer-list.nix` should come first

Otherwise, LGTM I think

norpol commented:
Hey @diogotcorreia I've rebased and incoroporated your suggestions.

* webmail was already upgraded (I'm not sure if the webmail version on nixpkgs master is compatible with the previous stalwart-mail release that wasn't bumped yet
* I've dropped `spam-filter.auto-update = lib.mkDefault false;` again as you've pointed out is already a default and if you just enable this option through the config users would see this warning https://stalw.art/docs/configuration/overview/#local-and-database-settings - maybe there are some things for future improvements to the modules that would optionally automatically expand the `config.local-keys` or maybe these options are most advisable managed through something else?
* commits are squashed, reorganized, ...

diogotcorreia reviewed (approved):
Sorry it took so long, I was busy for the past few days
LGTM!

norpol commented:
Thanks @diogotcorreia. I've created a separate PR for #425489 in the meanwhile.
